registry/client/repository.go

```diff
@@ -221,6 +240,10 @@ func (t *tags) All(ctx context.Context) ([]string, error) {
 		}
 		defer resp.Body.Close()
 
+		if resp.StatusCode == http.StatusNotFound && !compatURL {
```

Comment:
I think this needs to cover a broader range of possible status codes. Possibly any 4xx except 401. We haven't defined what the behavior must be for undefined routes which in some cases could leave proxies to set the rules.